[Quant][bc-breaking] Remove overwrite_output_observer

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88620

Summary: When the BackendConfig was first introduced,
`overwrite_output_observer` and `overwrite_output_fake_quantize`
were added to ensure fixed qparams ops like `torch.nn.Sigmoid`
and `torch.nn.Tanh` used the correct observers and fake quantizes.
However, this is hacky because the BackendConfig should not set
the observer constructors themselves, but should instead specify
only requirements on the observers.

Later, https://github.com/pytorch/pytorch/pull/80184 added the
correct observers to `get_default_qconfig_mapping` along with
validation logic that throws an error if incorrect observers
were specified. With this change, we no longer need to overwrite
the observers from the BackendConfig, since we expect the user to
pass in the correct observers for these ops.

This commit removes these overwrite observer settings in the
BackendConfig. Instead, we represent the observer constraints for
fixed qparams ops through the existing DTypeWithConstraints
mechanism. Note that, however, to be consistent with other
DTypeWithConstraints checks, we no longer throw an error if an
incorrect observer is specified, but simply ignore the offending
QConfig and log a warning instead. This is the BC-breaking part
of the change.

BC-breaking notes:

```
from torch.ao.quantization.qconfig import default_qconfig
from torch.ao.quantization.quantize_fx import prepare_fx

model = ModelWithFixedQParamsOps()
qconfig_mapping = QConfigMapping().set_global(default_qconfig)
example_inputs = ...
prepare_fx(model, qconfig_mapping, example_inputs)
```

Before this commit, running the above leads to an exception
because the wrong observers are used for fixed qparams ops.
After this commit, the above will only encounter a warning,
and the fixed qparams ops will not be quantized. In both cases,
switching to `get_default_qconfig_mapping` will cause the
fixed qparams ops to be quantized.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel

andrewor14 committed:
[Quant][bc-breaking] Remove overwrite_output_observer

Summary: When the BackendConfig was first introduced,
`overwrite_output_observer` and `overwrite_output_fake_quantize`
were added to ensure fixed qparams ops like `torch.nn.Sigmoid`
and `torch.nn.Tanh` used the correct observers and fake quantizes.
However, this is hacky because the BackendConfig should only
specify requirements on observers, but not the observer
constructors themselves.

Later, https://github.com/pytorch/pytorch/pull/80184 added the
correct observers to `get_default_qconfig_mapping` along with
validation logic that throws an error if incorrect observers
were specified. With this change, we no longer need to overwrite
the observers specified by the user.

This commit removes these observer settings in the BackendConfig.
Instead, we represent the observer constraints for fixed qparams
ops through the existing DTypeWithConstraints mechanism. To be
consistent with other DTypeWithConstraints checks, we no longer
throw an error if an incorrect observer is specified, but simply
ignore the offending QConfig and log a warning instead. This is
the BC-breaking part of the change.

BC-breaking notes:

```
from torch.ao.quantization.qconfig import default_qconfig
from torch.ao.quantization.quantize_fx import prepare_fx

model = ModelWithFixedQParamsOps()
qconfig_mapping = QConfigMapping().set_global(default_qconfig)
example_inputs = ...
prepare_fx(model, qconfig_mapping, example_inputs)
```

Before this commit, running the above leads to an exception
because the wrong observers are used for fixed qparams ops.
After this commit, the above will only encounter a warning,
and the fixed qparams ops will not be quantized. In both cases,
switching to `get_default_qconfig_mapping` will cause the
fixed qparams ops to be quantized.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][bc-breaking] Remove overwrite_output_observer"


Summary: When the BackendConfig was first introduced,
`overwrite_output_observer` and `overwrite_output_fake_quantize`
were added to ensure fixed qparams ops like `torch.nn.Sigmoid`
and `torch.nn.Tanh` used the correct observers and fake quantizes.
However, this is hacky because the BackendConfig should only
specify requirements on observers, but not the observer
constructors themselves.

Later, https://github.com/pytorch/pytorch/pull/80184 added the
correct observers to `get_default_qconfig_mapping` along with
validation logic that throws an error if incorrect observers
were specified. With this change, we no longer need to overwrite
the observers specified by the user.

This commit removes these observer settings in the BackendConfig.
Instead, we represent the observer constraints for fixed qparams
ops through the existing DTypeWithConstraints mechanism. To be
consistent with other DTypeWithConstraints checks, we no longer
throw an error if an incorrect observer is specified, but simply
ignore the offending QConfig and log a warning instead. This is
the BC-breaking part of the change.

BC-breaking notes:

```
from torch.ao.quantization.qconfig import default_qconfig
from torch.ao.quantization.quantize_fx import prepare_fx

model = ModelWithFixedQParamsOps()
qconfig_mapping = QConfigMapping().set_global(default_qconfig)
example_inputs = ...
prepare_fx(model, qconfig_mapping, example_inputs)
```

Before this commit, running the above leads to an exception
because the wrong observers are used for fixed qparams ops.
After this commit, the above will only encounter a warning,
and the fixed qparams ops will not be quantized. In both cases,
switching to `get_default_qconfig_mapping` will cause the
fixed qparams ops to be quantized.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][bc-breaking] Remove overwrite_output_observer"


Summary: When the BackendConfig was first introduced,
`overwrite_output_observer` and `overwrite_output_fake_quantize`
were added to ensure fixed qparams ops like `torch.nn.Sigmoid`
and `torch.nn.Tanh` used the correct observers and fake quantizes.
However, this is hacky because the BackendConfig should not set
the observer constructors themselves, but should instead specify
only requirements on the observers.

Later, https://github.com/pytorch/pytorch/pull/80184 added the
correct observers to `get_default_qconfig_mapping` along with
validation logic that throws an error if incorrect observers
were specified. With this change, we no longer need to overwrite
the observers from the BackendConfig, since we expect the user to
pass in the correct observers for these ops.

This commit removes these overwrite observer settings in the
BackendConfig. Instead, we represent the observer constraints for
fixed qparams ops through the existing DTypeWithConstraints
mechanism. Note that, however, to be consistent with other
DTypeWithConstraints checks, we no longer throw an error if an
incorrect observer is specified, but simply ignore the offending
QConfig and log a warning instead. This is the BC-breaking part
of the change.

BC-breaking notes:

```
from torch.ao.quantization.qconfig import default_qconfig
from torch.ao.quantization.quantize_fx import prepare_fx

model = ModelWithFixedQParamsOps()
qconfig_mapping = QConfigMapping().set_global(default_qconfig)
example_inputs = ...
prepare_fx(model, qconfig_mapping, example_inputs)
```

Before this commit, running the above leads to an exception
because the wrong observers are used for fixed qparams ops.
After this commit, the above will only encounter a warning,
and the fixed qparams ops will not be quantized. In both cases,
switching to `get_default_qconfig_mapping` will cause the
fixed qparams ops to be quantized.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

[ghstack-poisoned]

vkuzo commented:
> Note that, however, to be consistent with other
DTypeWithConstraints checks, we no longer throw an error if an
incorrect observer is specified, but simply ignore the offending
QConfig and log a warning instead.

Not sure I understand this yet.  Could you provide a specific example?  As a user, I would either expect things to work as I configured or to see an exception thrown, a warning and behavior change seems surprising

vkuzo commented on torch/ao/quantization/backend_config/_common_operator_config_utils.py:
```diff
@@ -393,21 +394,43 @@ def _get_default_op_configs(dtype_configs: List[DTypeConfig]) -> List[BackendPat
 def _get_fixed_qparams_op_configs(dtype_configs: List[DTypeConfig]) -> List[BackendPatternConfig]:
     fixed_qparams_op_configs = []
-    for fixed_qparam_op, output_observer in _FIXED_QPARAMS_OP_TO_OBSERVER.items():
+    for fixed_qparam_op, obs_ctr in _FIXED_QPARAMS_OP_TO_OBSERVER.items():
```
it's surprising to depend on qconfig here, since qconfig and backend_config are supposed to be configuring different things.

How about something like:
1. there is a mapping of PyTorch op to fixed qparams (not related to qconfig or backend_config)
2. there is a mapping from (1) to qconfig
3. there is a mapping from (1) to backend_config

andrewor14 replied:
Actually `_FIXED_QPARAMS_OP_TO_OBSERVER` is not part of qconfig, so what you're describing is already the current behavior. We do have a future BE task planned to move this map somewhere else (outside of qconfig.py), but that's separate from this PR. I can add a TODO here

vkuzo replied:
observer objects are intended to be used in `QConfig` only, so the current code is using qconfig related objects indirectly.  Can we fix this in the same stack as this PR?  It's confusing to reuse these abstractions outside of qconfig.

andrewor14 replied:
I see. Ok, what I can do then is make a map from the ops to the expected qparams, then derive the observers based on this map.

vkuzo commented on torch/ao/quantization/backend_config/backend_config.py:
```diff
@@ -69,14 +66,17 @@ class ObservationType(Enum):
     quant_max_upper_bound: Union[int, float, None] = None
     scale_min_lower_bound: Union[int, float, None] = None
     scale_max_upper_bound: Union[int, float, None] = None
+    fixed_scale: Optional[float] = None
```
how about something like `scale_exact_match` and `zero_point_exact_match`, to be consistent with the naming of the other attributes?

vkuzo commented:
> Sure. Today we check whether a qconfig satisfies the dtype constraints set by the backend [here](https://github.com/pytorch/pytorch/blob/4b898a7304246275b250b159dd0ac8e68a6df95d/torch/ao/quantization/fx/utils.py#L944) when we insert observers. The contract is if the qconfig doesn't satisfy these constraints, then we will simply log a warning and not insert the observers. We do this today for quant range and scale range constraints.

If I'm a user and I specify a qconfig with some constraints, I would expect for my configuration to either be honored, or to see an exception.  I wouldn't expect to have to look at warnings to see if the framework did something different from what I asked.  Can we just always throw an exception and remove the "silently change behavior and log a warning" cases?

andrewor14 commented:
> If I'm a user and I specify a qconfig with some constraints, I would expect for my configuration to either be honored, or to see an exception. I wouldn't expect to have to look at warnings to see if the framework did something different from what I asked. Can we just always throw an exception and remove the "silently change behavior and log a warning" cases?

OK, that seems like a broader change, since it also changes the existing behavior for other dtype constraints added. I think that was also the original proposal in https://github.com/pytorch/pytorch/pull/85200 but we ended up deciding to do the warning thing instead. Maybe we should first align on how dtype constraints should be handled. @jerryzh168 Should we change the checks [here](https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/utils.py#L944) to throw exceptions?

jerryzh168 commented:
I think it comes down to how we view the relationship of backend config and qconfig_mapping, we can start with a simple example, e.g. QConfigMapping().set_global(default_qconfig), this means: user is requesting to quantize the whole model with default_qconfig, but today how we handle this is we look at what is the supported config in backend_config and only quantize the operators that supports default_qconfig.

Error out when the requested operator + qconfig is not supported in backend_config would be a big change in semantics and might break a lot of models I think, I feel there will be operators in the model that's not in backend_config, and we also don't want to quantize them. Maybe we can check a few models to see if this would make sense or not first?

vkuzo commented:
> we look at what is the supported config in backend_config and only quantize the operators that supports default_qconfig.

I think this is reasonable if it's applied consistently.  Would we feel okay with removing the exception thrown from `prepare.py::_validate_fixed_qparams_qconfigs` then, to ensure this works the same way for all ops?

jerryzh168 commented:
In summary, let's say we have
qconfig_mapping: specifies (op, qconfig)
backend_config: have config entry for (op, dtype_constraints)

Current Behavior:
case 1 (op, qconfig) in qconfig_mapping matches (op, dtype_constraints) : quantize
case 2 (op, qconfig) in qconfig_mapping doesn't match (op, dtype_constraints), give warning for some and ignore some others (ideally I think we should give warning for everything), error for fixed qparam
case 3 (op, qconfig) in qconfig_mapping, but no op configuration in backend_config, ignore

Proposed
case 1 (op, qconfig) in qconfig_mapping matches (op, dtype_constraints) : quantize
case 2 (op, qconfig) in qconfig_mapping doesn't match (op, dtype_constraints), error
case 3 (op, qconfig) in qconfig_mapping, but no op configuration in backend_config, error

I'm afraid that there will be many ops in 2.3 and it will break a lot of ops

and I agree that it's better to make the second one consistent, I'm OK with warning/ignore/error, we can also discuss which one might make more sense.

jerryzh168 commented:
> > we look at what is the supported config in backend_config and only quantize the operators that supports default_qconfig.
> 
> I think this is reasonable if it's applied consistently. Would we feel okay with removing the exception thrown from `prepare.py::_validate_fixed_qparams_qconfigs` then, to ensure this works the same way for all ops?

yeah agree we should do this consistently, I think we can remove exception for now to be consistent and this is also what this PR is trying to do?

vkuzo commented:
I think we should document both https://pytorch.org/docs/master/generated/torch.ao.quantization.qconfig_mapping.QConfigMapping.html#torch.ao.quantization.qconfig_mapping.QConfigMapping and https://pytorch.org/docs/master/generated/torch.ao.quantization.backend_config.BackendConfig.html#torch.ao.quantization.backend_config.BackendConfig to clarify the behavior.  I think what a user would assume is "honor or exception", let's clearly explain if this isn't the case.

Not quantizing if qconfig + backend_config don't align sounds fine.  I think we should do the same thing for all ops, it's surprising to vary the behavior depending on the op.

andrewor14 commented:
Ok, that sounds good. Just to confirm, what we decided on then is the following behavior, correct?

case 1 (op, qconfig) in qconfig_mapping matches (op, dtype_constraints): quantize
case 2 (op, qconfig) in qconfig_mapping doesn't match (op, dtype_constraints): ignore (+ warning)
case 3 (op, qconfig) in qconfig_mapping, but no op configuration in backend_config: ignore

I will add some more documentation to QConfigMapping and BackendConfig as well to clarify this relationship

vkuzo commented:
> case 1 (op, qconfig) in qconfig_mapping matches (op, dtype_constraints): quantize
> case 2 (op, qconfig) in qconfig_mapping doesn't match (op, dtype_constraints): ignore (+ warning)
> case 3 (op, qconfig) in qconfig_mapping, but no op configuration in backend_config: ignore

How about not separating case 2 and case 3?  It doesn't have to be in this PR.  I think a warning which is off by default but that the user could turn on to debug why things aren't getting quantized could make sense here.  We already have a lot of log spew from quantization workflow APIs, so I don't think we should expect the user to pay attention to warnings unless they are actively trying to debug something.

andrewor14 committed:
Update on "[Quant][bc-breaking] Remove overwrite_output_observer"


Summary: When the BackendConfig was first introduced,
`overwrite_output_observer` and `overwrite_output_fake_quantize`
were added to ensure fixed qparams ops like `torch.nn.Sigmoid`
and `torch.nn.Tanh` used the correct observers and fake quantizes.
However, this is hacky because the BackendConfig should not set
the observer constructors themselves, but should instead specify
only requirements on the observers.

Later, https://github.com/pytorch/pytorch/pull/80184 added the
correct observers to `get_default_qconfig_mapping` along with
validation logic that throws an error if incorrect observers
were specified. With this change, we no longer need to overwrite
the observers from the BackendConfig, since we expect the user to
pass in the correct observers for these ops.

This commit removes these overwrite observer settings in the
BackendConfig. Instead, we represent the observer constraints for
fixed qparams ops through the existing DTypeWithConstraints
mechanism. Note that, however, to be consistent with other
DTypeWithConstraints checks, we no longer throw an error if an
incorrect observer is specified, but simply ignore the offending
QConfig and log a warning instead. This is the BC-breaking part
of the change.

BC-breaking notes:

```
from torch.ao.quantization.qconfig import default_qconfig
from torch.ao.quantization.quantize_fx import prepare_fx

model = ModelWithFixedQParamsOps()
qconfig_mapping = QConfigMapping().set_global(default_qconfig)
example_inputs = ...
prepare_fx(model, qconfig_mapping, example_inputs)
```

Before this commit, running the above leads to an exception
because the wrong observers are used for fixed qparams ops.
After this commit, the above will only encounter a warning,
and the fixed qparams ops will not be quantized. In both cases,
switching to `get_default_qconfig_mapping` will cause the
fixed qparams ops to be quantized.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][bc-breaking] Remove overwrite_output_observer"


Summary: When the BackendConfig was first introduced,
`overwrite_output_observer` and `overwrite_output_fake_quantize`
were added to ensure fixed qparams ops like `torch.nn.Sigmoid`
and `torch.nn.Tanh` used the correct observers and fake quantizes.
However, this is hacky because the BackendConfig should not set
the observer constructors themselves, but should instead specify
only requirements on the observers.

Later, https://github.com/pytorch/pytorch/pull/80184 added the
correct observers to `get_default_qconfig_mapping` along with
validation logic that throws an error if incorrect observers
were specified. With this change, we no longer need to overwrite
the observers from the BackendConfig, since we expect the user to
pass in the correct observers for these ops.

This commit removes these overwrite observer settings in the
BackendConfig. Instead, we represent the observer constraints for
fixed qparams ops through the existing DTypeWithConstraints
mechanism. Note that, however, to be consistent with other
DTypeWithConstraints checks, we no longer throw an error if an
incorrect observer is specified, but simply ignore the offending
QConfig and log a warning instead. This is the BC-breaking part
of the change.

BC-breaking notes:

```
from torch.ao.quantization.qconfig import default_qconfig
from torch.ao.quantization.quantize_fx import prepare_fx

model = ModelWithFixedQParamsOps()
qconfig_mapping = QConfigMapping().set_global(default_qconfig)
example_inputs = ...
prepare_fx(model, qconfig_mapping, example_inputs)
```

Before this commit, running the above leads to an exception
because the wrong observers are used for fixed qparams ops.
After this commit, the above will only encounter a warning,
and the fixed qparams ops will not be quantized. In both cases,
switching to `get_default_qconfig_mapping` will cause the
fixed qparams ops to be quantized.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][bc-breaking] Remove overwrite_output_observer"


Summary: When the BackendConfig was first introduced,
`overwrite_output_observer` and `overwrite_output_fake_quantize`
were added to ensure fixed qparams ops like `torch.nn.Sigmoid`
and `torch.nn.Tanh` used the correct observers and fake quantizes.
However, this is hacky because the BackendConfig should not set
the observer constructors themselves, but should instead specify
only requirements on the observers.

Later, https://github.com/pytorch/pytorch/pull/80184 added the
correct observers to `get_default_qconfig_mapping` along with
validation logic that throws an error if incorrect observers
were specified. With this change, we no longer need to overwrite
the observers from the BackendConfig, since we expect the user to
pass in the correct observers for these ops.

This commit removes these overwrite observer settings in the
BackendConfig. Instead, we represent the observer constraints for
fixed qparams ops through the existing DTypeWithConstraints
mechanism. Note that, however, to be consistent with other
DTypeWithConstraints checks, we no longer throw an error if an
incorrect observer is specified, but simply ignore the offending
QConfig and log a warning instead. This is the BC-breaking part
of the change.

BC-breaking notes:

```
from torch.ao.quantization.qconfig import default_qconfig
from torch.ao.quantization.quantize_fx import prepare_fx

model = ModelWithFixedQParamsOps()
qconfig_mapping = QConfigMapping().set_global(default_qconfig)
example_inputs = ...
prepare_fx(model, qconfig_mapping, example_inputs)
```

Before this commit, running the above leads to an exception
because the wrong observers are used for fixed qparams ops.
After this commit, the above will only encounter a warning,
and the fixed qparams ops will not be quantized. In both cases,
switching to `get_default_qconfig_mapping` will cause the
fixed qparams ops to be quantized.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/utils.py:
```diff
@@ -968,6 +979,8 @@ def _activation_post_process_satisfies_dtype_config_constraints(
         backend_quant_min = dtype_with_constraints.quant_min_lower_bound
         backend_quant_max = dtype_with_constraints.quant_max_upper_bound
         backend_scale_min = dtype_with_constraints.scale_min_lower_bound
+        backend_fixed_scale = dtype_with_constraints.scale_exact_match
+        backend_fixed_zero_point = dtype_with_constraints.zero_point_exact_match
```
nit: align the variable name

vkuzo commented on torch/ao/quantization/backend_config/README.md:
```diff
@@ -152,3 +152,7 @@ The user's QConfig may specify `quant_min` and `quant_max`, which are min and ma
+
+#### Fixed quantization parameters
+
+For ops with fixed quantization parameters such as `torch.nn.Sigmoid` or `torch.nn.Tanh`, the BackendConfig can specify the specific scale and zero point values as constraints on the input and output activations. The user's QConfigs for these ops must use `FixedQParamsObservers` or `FixedQParamsFakeQuantize` for their activations with matching scale and zero point values, otherwise these QConfigs ...
```
nit: `FixedQParamsObserver`?

andrewor14 committed:
Update on "[Quant][bc-breaking] Remove overwrite_output_observer"


Summary: When the BackendConfig was first introduced,
`overwrite_output_observer` and `overwrite_output_fake_quantize`
were added to ensure fixed qparams ops like `torch.nn.Sigmoid`
and `torch.nn.Tanh` used the correct observers and fake quantizes.
However, this is hacky because the BackendConfig should not set
the observer constructors themselves, but should instead specify
only requirements on the observers.

Later, https://github.com/pytorch/pytorch/pull/80184 added the
correct observers to `get_default_qconfig_mapping` along with
validation logic that throws an error if incorrect observers
were specified. With this change, we no longer need to overwrite
the observers from the BackendConfig, since we expect the user to
pass in the correct observers for these ops.

This commit removes these overwrite observer settings in the
BackendConfig. Instead, we represent the observer constraints for
fixed qparams ops through the existing DTypeWithConstraints
mechanism. Note that, however, to be consistent with other
DTypeWithConstraints checks, we no longer throw an error if an
incorrect observer is specified, but simply ignore the offending
QConfig and log a warning instead. This is the BC-breaking part
of the change.

BC-breaking notes:

```
from torch.ao.quantization.qconfig import default_qconfig
from torch.ao.quantization.quantize_fx import prepare_fx

model = ModelWithFixedQParamsOps()
qconfig_mapping = QConfigMapping().set_global(default_qconfig)
example_inputs = ...
prepare_fx(model, qconfig_mapping, example_inputs)
```

Before this commit, running the above leads to an exception
because the wrong observers are used for fixed qparams ops.
After this commit, the above will only encounter a warning,
and the fixed qparams ops will not be quantized. In both cases,
switching to `get_default_qconfig_mapping` will cause the
fixed qparams ops to be quantized.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -826,13 +819,7 @@ def maybe_insert_output_observer_for_node(
-                matched_pattern,
-                is_qat)
-        observer = act_post_process_ctr()
+        observer = qconfig.activation()
```
nice!

jerryzh168 commented on torch/ao/quantization/fx/utils.py:
```diff
@@ -990,6 +1006,30 @@ def _activation_post_process_satisfies_dtype_config_constraints(
+            )
+            if not isinstance(activation_post_process, FixedQParamsObserver) and \
+                    not isinstance(activation_post_process, FixedQParamsFakeQuantize):
+                warnings.warn(("QConfig must specify a FixedQParamsObserver or a FixedQParamsFakeQuantize "
```
just to confirm, in future PRs do we want to print warning for all branches that `return False`?

andrewor14 replied:
Yes exactly

jerryzh168 reviewed (approved):
LGTM, this makes our api cleaner and is also a great cleanup of the code base as well, thanks a lot

jerryzh168 commented on torch/ao/quantization/backend_config/qnnpack.py:
```diff
@@ -129,7 +129,6 @@ def get_qnnpack_backend_config() -> BackendConfig:
         qnnpack_default_op_quint8_dtype_config,
     ]
     fixed_qparams_op_dtype_configs = [
-        qnnpack_default_op_qint8_symmetric_dtype_config,
```
why this one is removed?

andrewor14 replied:
Because this dtype config wasn't compatible with the 0to1 and neg1to1 observer constraints (since these are in quint8), though it's technically not required for this PR and I can revert this change

andrewor14 committed:
Update on "[Quant][bc-breaking] Remove overwrite_output_observer"


Summary: When the BackendConfig was first introduced,
`overwrite_output_observer` and `overwrite_output_fake_quantize`
were added to ensure fixed qparams ops like `torch.nn.Sigmoid`
and `torch.nn.Tanh` used the correct observers and fake quantizes.
However, this is hacky because the BackendConfig should not set
the observer constructors themselves, but should instead specify
only requirements on the observers.

Later, https://github.com/pytorch/pytorch/pull/80184 added the
correct observers to `get_default_qconfig_mapping` along with
validation logic that throws an error if incorrect observers
were specified. With this change, we no longer need to overwrite
the observers from the BackendConfig, since we expect the user to
pass in the correct observers for these ops.

This commit removes these overwrite observer settings in the
BackendConfig. Instead, we represent the observer constraints for
fixed qparams ops through the existing DTypeWithConstraints
mechanism. Note that, however, to be consistent with other
DTypeWithConstraints checks, we no longer throw an error if an
incorrect observer is specified, but simply ignore the offending
QConfig and log a warning instead. This is the BC-breaking part
of the change.

BC-breaking notes:

```
from torch.ao.quantization.qconfig import default_qconfig
from torch.ao.quantization.quantize_fx import prepare_fx

model = ModelWithFixedQParamsOps()
qconfig_mapping = QConfigMapping().set_global(default_qconfig)
example_inputs = ...
prepare_fx(model, qconfig_mapping, example_inputs)
```

Before this commit, running the above leads to an exception
because the wrong observers are used for fixed qparams ops.
After this commit, the above will only encounter a warning,
and the fixed qparams ops will not be quantized. In both cases,
switching to `get_default_qconfig_mapping` will cause the
fixed qparams ops to be quantized.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

[ghstack-poisoned]

andrewor14 commented:
@pytorchbot merge
